Source/JavaScriptCore/runtime/WaiterListsManager.h

```diff
@@ -0,0 +1,330 @@
+    {
+        return queue.takeFirst(functor);
+    }
+
+    bool isEmpty()
+    {
+        return queue.isEmpty();
+    }
+
+    void dump(PrintStream& out) const
+    {
+        out.print("queue: [");
+        for (auto& waiter : queue) {
+            waiter->dump(out);
+            out.print(" ");
+        }
+        out.print("]");
+    }
+
+    mutable Mutex lock;
+
+private:
+    Deque<std::unique_ptr<Waiter>> queue;
+};
+
+class WaiterListsManager {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    static WaiterListsManager& singleton()
+    {
+        static LazyNeverDestroyed<WaiterListsManager> manager;
+        static std::once_flag onceKey;
+        std::call_once(onceKey, [&] {
+            manager.construct();
+        });
+        return manager;
+    }
+
+    void addWaiter(void* ptr, VM* vm)
+    {
+        WaiterList* queue = add(ptr);
+        RELEASE_ASSERT(queue);
+        {
+            MutexLocker locker(queue->lock);
+            queue->enqueue(vm);
+        }
+        m_registeredVMs.add(vm);
+        if (WaiterListsManagerInternal::verbose)
+            dataLogF("WaiterListsManager added a new Waiter to a waiterList for ptr %p\n", ptr);
+    }
+
+    void addWaiter(void* ptr, JSPromise* promise)
+    {
+        WaiterList* queue = add(ptr);
+        RELEASE_ASSERT(queue);
+        {
+            MutexLocker locker(queue->lock);
+            queue->enqueue(promise);
+        }
+        m_registeredVMs.add(&promise->globalObject()->vm());
+        if (WaiterListsManagerInternal::verbose)
+            dataLogF("WaiterListsManager added a new AsyncWaiter to a waiterList for ptr %p\n", ptr);
+    }
+
+    unsigned notifyWaiter(void* ptr, unsigned count)
+    {
+        RELEASE_ASSERT(ptr);
+        unsigned notified = 0;
+        WaiterList* queue = find(ptr);
+        if (!queue)
+            return notified;
+        {
+            MutexLocker locker(queue->lock);
+            while (notified < count) {
+                if (queue->isEmpty())
+                    break;
+                std::unique_ptr<Waiter> waiter = queue->dequeue();
+                bool isOKString = true;
+                if (notifyWaiterImpl(ptr, WTFMove(waiter), isOKString))
+                    notified++;
+            }
+        }
+        if (WaiterListsManagerInternal::verbose)
+            dataLogF("WaiterListsManager notified waiters (count %d) for ptr %p\n", notified, ptr);
+        return notified;
+    }
+
+    void timeoutAsyncWaiter(void* ptr, VM* vm)
+    {
+        RELEASE_ASSERT(ptr && vm);
+        WaiterList* queue = find(ptr);
+        RELEASE_ASSERT(queue);
+
+        {
+            MutexLocker locker(queue->lock);
+            std::unique_ptr<Waiter> waiter = queue->takeFirst([vm](auto& waiter) {
+                return waiter->isAsync() && waiter->getVM() == vm;
+            });
+            bool isOKString = false;
+            notifyWaiterImpl(ptr, WTFMove(waiter), isOKString);
```

Comment:
We should pass locker as an argument `AbstractLockHolder&` to represent that this function can be called only when we have a lock.